src/HTTPMultipartBodyParser.cpp

```diff
@@ -0,0 +1,278 @@
+#include "HTTPMultipartBodyParser.hpp"
+#include <sstream>
+
+const size_t MAXLINESIZE = 256;
+
+namespace httpsserver {
+
+HTTPMultipartBodyParser::HTTPMultipartBodyParser(HTTPRequest * req):
+  HTTPBodyParser(req),
+  peekBuffer(NULL),
+  peekBufferSize(0),
+  boundary(""),
+  lastBoundary(""),
+  fieldName(""),
+  fieldMimeType(""),
+  fieldFilename("")
+{
+  auto contentType = _request->getHeader("Content-Type");
+#ifdef DEBUG_MULTIPART_PARSER      
+  Serial.print("Content type: ");
+  Serial.println(contentType.c_str());
+#endif
+  auto boundaryIndex = contentType.find("boundary=");
+  if(boundaryIndex == std::string::npos) {
+    HTTPS_LOGE("Multipart: missing boundary=");
+    discardBody();
+    return;
+  }
+  //TODO: remove all magic constants
+  boundary = contentType.substr(boundaryIndex + 9);
+  auto commaIndex = boundary.find(';');
+  boundary = "--" + boundary.substr(0, commaIndex);
+  if(boundary.size() > 72) {
+    HTTPS_LOGE("Multipart: boundary string too long");
+    discardBody();
+  }
+  lastBoundary = boundary + "--";
+}
+
+HTTPMultipartBodyParser::~HTTPMultipartBodyParser() {
+  if (peekBuffer) {
+    free(peekBuffer);
+    peekBuffer = NULL;
+  }
+}
+
+void HTTPMultipartBodyParser::discardBody() {
+  if (peekBuffer) {
+    free(peekBuffer);
+  }
+  peekBuffer = NULL;
+  peekBufferSize = 0;
+  _request->discardRequestBody();
+}
+
+bool HTTPMultipartBodyParser::endOfBody() {
+  return peekBufferSize == 0  && _request->requestComplete();
+}
+
+void HTTPMultipartBodyParser::fillBuffer(size_t maxLen) {
+  // Fill the buffer with up to maxLen bytes (total length, including
+  // what was already in the buffer), but stop reading ahead once
+  // we have a CR in the buffer (because the upper layers will
+  // stop consuming there anyway, to forestall overrunning
+  // a boundary)
+  char *bufPtr;
+  if (peekBuffer == NULL) {
+    // Nothing in the buffer. Allocate one of the wanted size
+    peekBuffer = (char *)malloc(maxLen);
+    bufPtr = peekBuffer;
+    peekBufferSize = 0;
+  } else if (peekBufferSize < maxLen) {
+    // Something in the buffer, but not enough
+    peekBuffer = (char *)realloc(peekBuffer, maxLen);
```

Comment:
Handling of out-of-memory?

Also: `realloc` will not `free()` the memory in `peekBuffer` if it fails, so a different pointer should be used to store the result.